README.md
# pino-prettifier

[![npm version](https://badge.fury.io/js/pino-prettifier.svg)](http://badge.fury.io/js/pino-prettifier)

This package provides a log prettifier for the [Pino](https://getpino.io/) logging library.

## Usage

Install the library as usual:

```sh
npm install pino-prettifier
```

Then use the proper `pino` options:

```JS
const pino = require('pino');
const prettifier = require('pino-prettifier');
const logger = pino({
  prettyPrint: true,
  prettifier
});
```

## License

[MIT](LICENSE)


CHANGELOG.md
## Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

### v1.0.8 (2019-11-11)

- [`#69`](https://github.com/lifion/pino-prettifier/pull/69): Update dependency prettier to ^1.19.1
- [`#68`](https://github.com/lifion/pino-prettifier/pull/68): Update dependency auto-changelog to ^1.16.2
- [`#67`](https://github.com/lifion/pino-prettifier/pull/67): Update dependency eslint to ^6.6.0
- [`1081285`](https://github.com/lifion/pino-prettifier/commit/10812855dbc35744842ee0f1814f555c4880e455): Adopt core commons 2.3.4

### v1.0.7 (2019-10-21)

- [`#61`](https://github.com/lifion/pino-prettifier/pull/61): Update dependency codecov to ^3.6.1
- [`#62`](https://github.com/lifion/pino-prettifier/pull/62): Update dependency lint-staged to ^9.3.0
- [`#60`](https://github.com/lifion/pino-prettifier/pull/60): Update dependency eslint to ^6.4.0
- [`#59`](https://github.com/lifion/pino-prettifier/pull/59): Update dependency auto-changelog to ^1.16.1
- [`#56`](https://github.com/lifion/pino-prettifier/pull/56): Update dependency eslint to ^6.3.0
- [`#57`](https://github.com/lifion/pino-prettifier/pull/57): Update dependency auto-changelog to ^1.16.0
- [`#58`](https://github.com/lifion/pino-prettifier/pull/58): Update dependency husky to ^3.0.5
- [`#55`](https://github.com/lifion/pino-prettifier/pull/55): Update dependency lint-staged to ^9.2.5
- [`#48`](https://github.com/lifion/pino-prettifier/pull/48): Update dependency husky to ^3.0.2
- [`#47`](https://github.com/lifion/pino-prettifier/pull/47): Update dependency lint-staged to ^9.2.1
- [`#46`](https://github.com/lifion/pino-prettifier/pull/46): Update dependency husky to ^3.0.1
- [`#45`](https://github.com/lifion/pino-prettifier/pull/45): Update dependency eslint-config-lifion to ^1.2.6
- [`#40`](https://github.com/lifion/pino-prettifier/pull/40): Update dependency lint-staged to v9
- [`#41`](https://github.com/lifion/pino-prettifier/pull/41): Update dependency husky to v3
- [`8ec801e`](https://github.com/lifion/pino-prettifier/commit/8ec801e2bd2cc2e37c5f7a0873615733783a4182): Adopt core commons 2.3.3
- [`79f9505`](https://github.com/lifion/pino-prettifier/commit/79f9505cfce2aae0e9a67fae7e18139289aa352c): Upgrade dependencies
- [`90760ea`](https://github.com/lifion/pino-prettifier/commit/90760eacce1e6ec2d7abfe1a1e722ccb3b47c8ac): Downgrade ESLint
- [`85129f6`](https://github.com/lifion/pino-prettifier/commit/85129f6a60b7cdd1a5ba3de12b17051ec2fd89e4): Create CODE_OF_CONDUCT.md
- [`16f61a1`](https://github.com/lifion/pino-prettifier/commit/16f61a12d9f9394995a8a68a1f2140b777ce9e2c): Adopt core commons 2.3.3
- [`c6ae1a2`](https://github.com/lifion/pino-prettifier/commit/c6ae1a20ff6516131315311b0120f4930a02b03c): Update pull-request-check.yml

### v1.0.6 (2019-07-11)

- [`#39`](https://github.com/lifion/pino-prettifier/pull/39): Update dependency husky to ^2.7.0
- [`#38`](https://github.com/lifion/pino-prettifier/pull/38): Update dependency eslint to ^6.0.1
- [`#36`](https://github.com/lifion/pino-prettifier/pull/36): Update dependency eslint to v6
- [`#37`](https://github.com/lifion/pino-prettifier/pull/37): Update dependency husky to ^2.5.0
- [`29a7d33`](https://github.com/lifion/pino-prettifier/commit/29a7d332e0ab4750d4318fe675d2cf5117bdee8f): Adopt GitHub Actions
- [`63f60e0`](https://github.com/lifion/pino-prettifier/commit/63f60e07b7e47f7b0fcaa7925ff8999448e5d6dd): Update dependency husky to v3
- [`37e6a4f`](https://github.com/lifion/pino-prettifier/commit/37e6a4fd28f8cd5bb4169f3c4dac2da2d09f0350): Update main.workflow

### v1.0.5 (2019-06-19)

- [`#32`](https://github.com/lifion/pino-prettifier/pull/32): Update dependency prettier to ^1.18.2
- [`#33`](https://github.com/lifion/pino-prettifier/pull/33): Update dependency husky to ^2.4.1
- [`#34`](https://github.com/lifion/pino-prettifier/pull/34): Update dependency lint-staged to ^8.2.1
- [`#35`](https://github.com/lifion/pino-prettifier/pull/35): Update dependency eslint-config-lifion to ^1.2.3
- [`#30`](https://github.com/lifion/pino-prettifier/pull/30): Update dependency husky to ^2.4.0
- [`#31`](https://github.com/lifion/pino-prettifier/pull/31): Update dependency lint-staged to ^8.2.0
- [`#29`](https://github.com/lifion/pino-prettifier/pull/29): Bump handlebars from 4.1.1 to 4.1.2
- [`#28`](https://github.com/lifion/pino-prettifier/pull/28): Update dependency eslint-config-lifion to ^1.2.2
- [`#27`](https://github.com/lifion/pino-prettifier/pull/27): Update dependency eslint-config-lifion to ^1.2.1
- [`#26`](https://github.com/lifion/pino-prettifier/pull/26): Update dependency lint-staged to ^8.1.7
- [`#25`](https://github.com/lifion/pino-prettifier/pull/25): Update dependency codecov to ^3.5.0
- [`#24`](https://github.com/lifion/pino-prettifier/pull/24): Update dependency husky to ^2.3.0
- [`#23`](https://github.com/lifion/pino-prettifier/pull/23): Update dependency prettier to ^1.17.1
- [`#22`](https://github.com/lifion/pino-prettifier/pull/22): Update dependency codecov to ^3.4.0
- [`#20`](https://github.com/lifion/pino-prettifier/pull/20): Update dependency lint-staged to ^8.1.6
- [`#21`](https://github.com/lifion/pino-prettifier/pull/21): Update dependency jest to ^24.8.0
- [`#19`](https://github.com/lifion/pino-prettifier/pull/19): Update dependency jest-junit to ^6.4.0
- [`#18`](https://github.com/lifion/pino-prettifier/pull/18): Update dependency husky to ^2.2.0
- [`#17`](https://github.com/lifion/pino-prettifier/pull/17): Update dependency husky to v2
- [`#16`](https://github.com/lifion/pino-prettifier/pull/16): Update dependency auto-changelog to ^1.13.0
- [`#15`](https://github.com/lifion/pino-prettifier/pull/15): Update dependency prettier to ^1.17.0

lib/index.test.js
/* eslint-disable no-control-regex */

'use strict';

const index = require('.');

describe('lib/index', () => {
  const logger = index();

  test('the module exports the expected', () => {
    expect(index).toEqual(expect.any(Function));
  });

  test('the module exports a function factory', () => {
    expect(index({})).toEqual(expect.any(Function));
  });

  test('the prettifier works with no configuration nor arguments', () => {
    expect(logger()).toMatch(/^\u001B\[90m\u001B\[39m USERLVL <.+> \[.+:\d+\]\n\s{14}$/);
  });

  test('time in numerical form is properly formatted', () => {
    const output = logger({ time: new Date().getTime() });
    expect(output).toMatch(/^\u001B\[90m\d{2}:\d{2}:\d{2}\.\d{3}\u001B\[39m/);
  });

  test('time in a quoted form is properly formatted', () => {
    const output = logger({ time: JSON.stringify(new Date()) });
    expect(output).toMatch(/^\u001B\[90m\d{2}:\d{2}:\d{2}\.\d{3}\u001B\[39m/);
  });

  test('debug messages are properly formatted', () => {
    const output = logger({ level: 20 });
    expect(output).toMatch(/\u001B\[34mDEBUG <.+> \[.+:\d+\]\u001B\[39m\n/);
  });

  test('error messages are properly formatted', () => {
    const output = logger({ level: 50 });
    expect(output).toMatch(/\u001B\[31mERROR <.+> \[.+:\d+\]\u001B\[39m\n/);
  });

  test('warning messages are properly formatted', () => {
    const output = logger({ level: 40 });
    expect(output).toMatch(/\u001B\[33mWARN <.+> \[.+:\d+\]\u001B\[39m\n/);
  });

  test('JSON fragments are properly formatted', () => {
    const output = logger({ msg: `${JSON.stringify({ foo: 'bar' })} baz` });
    expect(output).toMatch("{ foo: \u001B[32m'bar'\u001B[39m } baz\n");
  });

  test('pino stack lines are ignored', () => {
    const orgCaptureStack = Error.captureStackTrace;
    Error.captureStackTrace = obj => {
      Object.assign(obj, {
        stack: [
          ' at Object.write (/home/foo/node_modules/pino/lib/tools.js:1:2)',
          ' at Pino.write (/home/foo/node_modules/pino/lib/proto.js:3:4)',
          ' at bar (baz:5:6)',
          ' at qux (quux:7:8)'
        ].join('\n')
      });
    };
    expect(logger()).toMatch('<qux> [quux:7]');
    Error.captureStackTrace = orgCaptureStack;
  });

  test('incomplete stack lines are ignored', () => {
    const orgCaptureStack = Error.captureStackTrace;
    Error.captureStackTrace = obj => {
      Object.assign(obj, {
        stack: [
          ' at Object.write (/home/foo/node_modules/pino/lib/tools.js:1:2)',
          ' at Pino.write (/home/foo/node_modules/pino/lib/proto.js:3:4)',
          ' at bar (baz:5:6)'
        ].join('\n')
      });
    };
    expect(logger()).toMatch('USERLVL \n');
    Error.captureStackTrace = orgCaptureStack;
  });

  test('array messages are properly formatted', () => {
    const output = logger({ msg: [{ foo: 'bar' }] });
    expect(output).toMatch("[ { foo: \u001B[32m'bar'\u001B[39m } ]\n");
  });

  test('array fragments are properly formatted', () => {
    const output = logger({ msg: `${JSON.stringify([{ foo: 'bar' }])} baz` });
    expect(output).toMatch("[ { foo: \u001B[32m'bar'\u001B[39m } ] baz\n");
  });
});


lib/utils.test.js
/* eslint-disable no-control-regex */

'use strict';

const utils = require('./utils');

describe('lib/utils', () => {
  test('the module exports the expected', () => {
    expect(utils).toEqual({
      colors: expect.any(Object),
      padNum: expect.any(Function)
    });
  });

  test('the colors export are all functions that return colored-strings', () => {
    const { colors } = utils;
    Object.keys(colors).forEach(key => {
      const color = colors[key];
      expect(color).toEqual(expect.any(Function));
      expect(color('foo')).toMatch(/^\u001B\[\d+mfoo\u001B\[39m$/);
    });
  });

  test('the padNum export returns the expected formatted decimal number', () => {
    const { padNum } = utils;
    expect(padNum(12.345, 5)).toBe('00012');
  });
});


lib/index.js
'use strict';

const { inspect } = require('util');
const { relative } = require('path');
const { colors, padNum } = require('./utils');

const levels = {
  10: 'TRACE',
  20: 'DEBUG',
  30: 'INFO',
  40: 'WARN',
  50: 'ERROR',
  60: 'FATAL'
};

const exportsRegex = /^(Object\.)?module\.exports\./;
const sourceRegex = /^\s+at (.+) \((.+):(\d+):\d+/;

const INDENT = ' '.repeat(13);
const CWD = process.cwd();

const formatTime = time => {
  if (!time) return time;
  const epoch = !Number.isInteger(time) ? Date.parse(JSON.parse(time)) : time;
  const epochS = epoch / 1000;
  const epochM = epochS / 60;
  const ms = padNum(epoch % 1000, 3);
  const s = padNum(epochS % 60, 2);
  const m = padNum(epochM % 60, 2);
  const h = padNum((epochM / 60) % 24, 2);
  return [h, ':', m, ':', s, '.', ms].join('');
};

/**
 * Returns the corresponding colorization function for the provided logging level.
 *
 * @param {number} level - The logging level.
 * @returns {Function}
 */
const getLevelColor = level => {
  switch (level) {
    case 20:
      return colors.blue;
    case 50:
      return colors.red;
    case 40:
      return colors.yellow;
    default:
      return str => str;
  }
};

const formatMsg = (msg = '') => {
  return (typeof msg === 'string'
    ? msg
        .split(' ')
        .map(chunk => {
          if (
            (chunk.startsWith('{"') && chunk.endsWith('}')) ||
            (chunk.startsWith('[{"') && chunk.endsWith('}]'))
          ) {
            return inspect(JSON.parse(chunk), { colors: true });
          }
          return chunk;
        })
        .join(' ')
    : inspect(msg, { colors: true })
  )
    .split('\n')
    .map(i => [INDENT, i].join(''))
    .join('\n');
};

const getSource = stack => {
  let stackLines = stack.split('\n').slice(1);
  const firstNonPinoIndex = stackLines.findIndex(i => !i.includes('/node_modules/pino/'));
  stackLines = stackLines.slice(firstNonPinoIndex + 1);
  const [stackFrame] = stackLines;
  const stackData = stackFrame && stackFrame.match(sourceRegex);
  if (!stackData) return '';
  const [, at, filename, lineNumber] = stackData;
  const functionName = at.replace(exportsRegex, '');
  const relativePath = relative(CWD, filename);
  return ['<', functionName, '> [', relativePath, ':', lineNumber, ']'].join('');
};

module.exports = ({ messageKey = 'msg' } = {}) => {
  /**
   * A prettifier.
   *
   * @param {object} [data] - The data to prettify.
   * @param {number} [data.level] - The logging level.
   * @param {string|number} [data.time] - The timestamp.
   * @returns {string}
   */
  function prettifier(data = {}) {
    const captureObj = {};
    Error.captureStackTrace(captureObj, prettifier);
    const source = getSource(captureObj.stack);
    const msg = data[messageKey];
    const { level, time } = data;
    const prefix = colors.gray([formatTime(time)].join(''));
    const color = getLevelColor(level);
    const text = color([levels[level] || 'USERLVL', ' ', source].join(''));
    return [prefix, ' ', text, '\n', formatMsg(msg), '\n'].join('');
  }
  return prettifier;
};
